keps/sig-scheduling/3101-pod-to-node-assignment/README.md

```diff
@@ -0,0 +1,872 @@
+determine graduation.
+
+Consider the following in developing the graduation criteria for this enhancement:
+- [Maturity levels (`alpha`, `beta`, `stable`)][maturity-levels]
+- [Deprecation policy][deprecation-policy]
+
+Clearly define what graduation means by either linking to the [API doc
+definition](https://kubernetes.io/docs/concepts/overview/kubernetes-api/#api-versioning)
+or by redefining what graduation means.
+
+In general we try to use the same stages (alpha, beta, GA), regardless of how the
+functionality is accessed.
+
+[maturity-levels]: https://git.k8s.io/community/contributors/devel/sig-architecture/api_changes.md#alpha-beta-and-stable-versions
+[deprecation-policy]: https://kubernetes.io/docs/reference/using-api/deprecation-policy/
+
+Below are some examples to consider, in addition to the aforementioned [maturity levels][maturity-levels].
+
+#### Alpha
+
+- Feature implemented behind a feature flag
+- Initial e2e tests completed and enabled
+
+#### Beta
+
+- Gather feedback from developers and surveys
+- Complete features A, B, C
+- Additional tests are in Testgrid and linked in KEP
+
+#### GA
+
+- N examples of real-world usage
+- N installs
+- More rigorous forms of testing—e.g., downgrade tests and scalability tests
+- Allowing time for feedback
+
+**Note:** Generally we also wait at least two releases between beta and
+GA/stable, because there's no opportunity for user feedback, or even bug reports,
+in back-to-back releases.
+
+**For non-optional features moving to GA, the graduation criteria must include
+[conformance tests].**
+
+[conformance tests]: https://git.k8s.io/community/contributors/devel/sig-architecture/conformance-tests.md
+
+#### Deprecation
+
+- Announce deprecation and support policy of the existing flag
+- Two versions passed since introducing the functionality that deprecates the flag (to address version skew)
+- Address feedback on usage/changed behavior, provided on GitHub issues
+- Deprecate the flag
+-->
+
+#### Alpha -> Beta Graduation
+
+-   Unit and integration tests passing.
+-   Feature implemented behind a feature flag.
+-   Benchmark tests showing no performance problems.
+-   No user complaints regarding performance/correctness.
+
+#### Beta -> GA Graduation
+
+-   Still no complaints regarding performance.
+
+### Upgrade / Downgrade Strategy
+
+<!--
+If applicable, how will the component be upgraded and downgraded? Make sure
+this is in the test plan.
+
+Consider the following in developing an upgrade/downgrade strategy for this
+enhancement:
+- What changes (in invocations, configurations, API use, etc.) is an existing
+  cluster required to make on upgrade, in order to maintain previous behavior?
+- What changes (in invocations, configurations, API use, etc.) is an existing
+  cluster required to make on upgrade, in order to make use of the enhancement?
+-->
+
+In the event of a downgrade, the `k8s.io/pod-per-node=true` annotation will be
+ignored even if it was set to true.
```

Comment:
Mention the following:
1. While the feature flag is disabled, the the field will be ignored by the api-server and will not be stored in etcd.
2. Not setting the new field will maintain the old behavior on both upgrade and downgrade of the control plane.
3. Setting the field after upgrading and while the feature flag is enabled will allow users to make use of the feature.
4. downgrading after upgrade or disabling the feature flag will 
  4.1 maintain previously set values in the field on existing pods, but they will have no effect. 
  4.2 disable the scheduler filter, and so the cluster may end up in a state where there are pods with `SingleExcludingDaemonsetAndStaticPods` previously assigned to nodes, and those nodes after the downgrade will now be allowed to fit "worker" pods. This situation will not be fixed after a subsequent upgrade, node drain is required to fix that.